File: apps/cli/Crypto/DekResolver.cs
using System.Security.Cryptography;
using DepVault.Cli.Auth;
using DepVault.Cli.Config;
using DepVault.Cli.Output;
using DepVault.Cli.Utils;
using Microsoft.Kiota.Abstractions;
using Spectre.Console;
using KeygrantBody = DepVault.Cli.ApiClient.Api.Projects.Item.Keygrants.KeygrantsPostRequestBody;
using GrantType = DepVault.Cli.ApiClient.Api.Projects.Item.Keygrants.KeygrantsPostRequestBody_grantType;

namespace DepVault.Cli.Crypto;

/// <summary>Resolves the Data Encryption Key (DEK) for client-side vault crypto.</summary>
public sealed class DekResolver(
    IApiClientFactory clientFactory,
    ICredentialStore credentialStore,
    IConsolePrompter prompter,
    CommandContext commandContext,
    VaultState vaultState)
{
    /// <summary>
    /// Collects the vault password interactively if needed. Returns null when the KEK is
    /// already cached or in CI mode. Must be called outside Spectre dynamic displays.
    /// </summary>
    public string? CollectVaultPassword()
    {
        if (commandContext.IsCiMode())
        {
            return null;
        }

        if (vaultState.IsUnlocked)
        {
            return null;
        }

        var password = Environment.GetEnvironmentVariable("DEPVAULT_PASSWORD");
        if (!string.IsNullOrEmpty(password))
        {
            return password;
        }

        if (!prompter.IsInteractive)
        {
            AnsiConsole.MarkupLine("[red]DEPVAULT_PASSWORD is required in non-interactive mode.[/]");
            return null;
        }

        return prompter.AskSecret("Enter vault password");
    }

    /// <summary>Resolves the DEK bytes by unwrapping from CI token or vault password.</summary>
    public async Task<byte[]?> ResolveAsync(string projectId, string? vaultPassword, CancellationToken ct)
    {
        return commandContext.IsCiMode()
            ? await ResolveCiDekAsync(ct)
            : await ResolveJwtDekAsync(projectId, vaultPassword, ct);
    }

    private async Task<byte[]?> ResolveCiDekAsync(CancellationToken ct)
    {
        var rawToken = Environment.GetEnvironmentVariable(Constants.CiTokenEnvVar)!;
        var client = clientFactory.Create();

        var ciSecrets = await client.Api.Ci.Secrets.GetAsync(cancellationToken: ct);

        if (ciSecrets is null)
        {
            AnsiConsole.MarkupLine("[red]Failed to fetch CI secrets metadata.[/]");
            return null;
        }

        var ciWrapKey = VaultCrypto.DeriveCiWrapKey(rawToken);
        return VaultCrypto.UnwrapKey(
            ciSecrets.WrappedDek ?? "", ciSecrets.WrappedDekIv ?? "",
            ciSecrets.WrappedDekTag ?? "", ciWrapKey);
    }

    private async Task<byte[]?> ResolveJwtDekAsync(string projectId, string? password, CancellationToken ct)
    {
        // 1. Check DEK cache first
        var cachedDek = vaultState.GetCachedDek(projectId);
        if (cachedDek is not null)
        {
            return cachedDek;
        }

        // 2. Fetch vault status. We always need it — to detect a salt rotation
        //    against a stale cached KEK, and to verify a freshly derived KEK.
        var token = credentialStore.Load()?.AccessToken;
        if (string.IsNullOrEmpty(token))
        {
            AnsiConsole.MarkupLine("[red]Not authenticated. Run 'depvault login' first.[/]");
            return null;
        }

        var apiClient = clientFactory.Create();
        var vaultStatus = await apiClient.Api.Vault.Status.GetAsync(cancellationToken: ct);

        if (vaultStatus is null || string.IsNullOrEmpty(vaultStatus.KekSalt))
        {
            AnsiConsole.MarkupLine("[red]Failed to fetch vault status. Ensure vault is initialized.[/]");
            return null;
        }

        // 3. Invalidate a cached KEK whose salt no longer matches the server's.
        //    This happens when the vault password was rotated from another client.
        if (vaultState.IsUnlocked && vaultState.KekSalt != vaultStatus.KekSalt)
        {
            AnsiConsole.MarkupLine(
                "[yellow]Vault password appears to have changed elsewhere — re-unlocking.[/]");
            vaultState.Lock();
        }

        // 4. Resolve KEK: use cached (now known-fresh) or derive from password
        byte[] kek;
        if (vaultState.IsUnlocked)
        {
            kek = vaultState.Kek!;
        }
        else
        {
            if (string.IsNullOrEmpty(password))
            {
                AnsiConsole.MarkupLine("[red]Vault password is required.[/]");
                return null;
            }

            var salt = Convert.FromBase64String(vaultStatus.KekSalt);
            var iterations = vaultStatus.KekIterations > 0 ? vaultStatus.KekIterations.Value : 600_000;
            kek = VaultCrypto.DeriveKek(password, salt, iterations);

            // Verify the KEK by unwrapping the server-stored wrapped private key.
            // Without this, a wrong password silently produces a new SELF grant
            // wrapped with an unusable KEK, corrupting future decrypts.
            if (!VerifyKek(kek, vaultStatus))
            {
                CryptographicOperations.ZeroMemory(kek);
                AnsiConsole.MarkupLine("[red]Incorrect vault password.[/]");
                return null;
            }

            vaultState.Unlock(kek, vaultStatus.KekSalt);
        }

        // 5. Fetch key grant and unwrap DEK
        try
        {
            var keyGrant = await apiClient.Api.Projects[projectId].Keygrants.My
                .GetAsync(cancellationToken: ct);

            if (keyGrant is not null && !string.IsNullOrEmpty(keyGrant.WrappedDek))
            {
                var dek = VaultCrypto.UnwrapKey(
                    keyGrant.WrappedDek, keyGrant.WrappedDekIv ?? "", keyGrant.WrappedDekTag ?? "", kek);
                vaultState.CacheDek(projectId, dek);
                return dek;
            }

            // No grant found — auto-create SELF grant
            return await CreateSelfGrantAsync(projectId, kek, ct);
        }
        catch (ApiException ex) when (ex.ResponseStatusCode is 404)
        {
            return await CreateSelfGrantAsync(projectId, kek, ct);
        }
    }

    /// <summary>Generate a new DEK, wrap with KEK, and POST a SELF grant to the server.</summary>
    private async Task<byte[]?> CreateSelfGrantAsync(string projectId, byte[] kek, CancellationToken ct)
    {
        var creds = credentialStore.Load();
        if (creds?.UserId is null)
        {
            PrintKeyGrantError();
            return null;
        }

        try
        {
            var dek = RandomNumberGenerator.GetBytes(32);
            var (wrappedDek, iv, tag) = VaultCrypto.WrapKey(dek, kek);

            var client = clientFactory.Create();
            await client.Api.Projects[projectId].Keygrants.PostAsync(new KeygrantBody
            {
                UserId = Guid.Parse(creds.UserId),
                WrappedDek = wrappedDek,
                WrappedDekIv = iv,
                WrappedDekTag = tag,
                GrantType = GrantType.SELF
            }, cancellationToken: ct);

            vaultState.CacheDek(projectId, dek);
            AnsiConsole.MarkupLine("[green]Created encryption key for this project.[/]");
            return dek;
        }
        catch
        {
            PrintKeyGrantError();
            return null;
        }
    }

    /// <summary>
    /// Verifies a candidate KEK by attempting to unwrap the server-stored wrapped private key.
    /// A successful unwrap proves the KEK was derived from the correct password and salt; a
    /// failed unwrap (AES-GCM auth tag mismatch) proves it was not. Returns true on success.
    /// </summary>
    private static bool VerifyKek(byte[] kek, ApiClient.Api.Vault.Status.StatusGetResponse status)
    {
        if (string.IsNullOrEmpty(status.WrappedPrivateKey) ||
            string.IsNullOrEmpty(status.WrappedPrivateKeyIv) ||
            string.IsNullOrEmpty(status.WrappedPrivateKeyTag))
        {
            // Vault has no wrapped private key yet — can't verify, but this shouldn't happen
            // for a fully-initialized vault. Allow through rather than block on bad server state.
            return true;
        }

        try
        {
            var raw = VaultCrypto.UnwrapKey(
                status.WrappedPrivateKey, status.WrappedPrivateKeyIv, status.WrappedPrivateKeyTag, kek);
            CryptographicOperations.ZeroMemory(raw);
            return true;
        }
        catch (CryptographicException)
        {
            return false;
        }
    }

    private static void PrintKeyGrantError()
    {
        AnsiConsole.WriteLine();
        AnsiConsole.Write(new Panel(
                new Rows(
                    new Markup("[red]No encryption key grant found for this project.[/]"),
                    new Markup(""),
                    new Markup("[grey]Key grants are created when you first open a project's vault[/]"),
                    new Markup("[grey]in the web dashboard. To fix this:[/]"),
                    new Markup(""),
                    new Markup("  [cyan1]1.[/] [grey]Open the DepVault web dashboard[/]"),
                    new Markup("  [cyan1]2.[/] [grey]Navigate to this project → [bold]Vault[/] tab[/]"),
                    new Markup("  [cyan1]3.[/] [grey]Unlock the vault with your password[/]"),
                    new Markup("  [cyan1]4.[/] [grey]Run this CLI command again[/]"),
                    new Markup(""),
                    new Markup("[grey]If you are a team member, ask the project owner to grant[/]"),
                    new Markup("[grey]you access from the [bold]Team[/] settings page.[/]")))
            .Header("[red]Key Grant Missing[/]")
            .Border(BoxBorder.Rounded)
            .BorderStyle(new Style(Color.Red))
            .Padding(1, 0));
        AnsiConsole.WriteLine();
    }
}

File: apps/cli-ts/src/services/dek-resolver.ts
import {
  deriveCIWrapKey,
  deriveKEK,
  exportDEK,
  fromBase64,
  generateDEK,
  importDEK,
  unwrapKey,
  wrapKey,
} from "@depvault/crypto";
import { CI_TOKEN_ENV_VAR, VAULT_PASSWORD_ENV_VAR } from "@/constants";
import { getApiClient } from "./api-client";
import { loadCredentials } from "./credentials";

/**
 * Resolves the project DEK (Data Encryption Key).
 * - CI mode: derives wrap key from DEPVAULT_TOKEN via HKDF, unwraps DEK from CI secrets endpoint
 * - JWT mode: derives KEK from vault password via PBKDF2, unwraps DEK from user's key grant
 */
export async function resolveDek(projectId: string, kek: CryptoKey | null): Promise<CryptoKey> {
  const ciToken = process.env[CI_TOKEN_ENV_VAR];

  if (ciToken) {
    return resolveDekFromCiToken(ciToken);
  }

  if (kek) {
    return resolveDekFromKek(projectId, kek);
  }

  // One-shot mode: derive KEK from vault password
  const password = process.env[VAULT_PASSWORD_ENV_VAR];
  if (!password) {
    throw new Error(
      `Vault is locked. Use /unlock in REPL mode, or set ${VAULT_PASSWORD_ENV_VAR} env var.`,
    );
  }

  const derivedKek = await deriveKekFromPassword(password);
  return resolveDekFromKek(projectId, derivedKek);
}

/** Derive KEK from a vault password by fetching the salt from the vault status endpoint. */
export async function deriveKekFromPassword(password: string): Promise<CryptoKey> {
  const client = getApiClient();
  const { data: status, error } = await client.api.vault.status.get();

  if (error) {
    const errorValue = error.value as { message?: string; code?: number } | undefined;
    if (errorValue?.code === 401 || errorValue?.code === 404) {
      throw new Error(
        "Authentication failed. Your session may have expired — run /login to re-authenticate.",
      );
    }
    throw new Error(`Failed to fetch vault status: ${errorValue?.message ?? "Unknown error"}`);
  }

  if (!status?.hasVault || !status.kekSalt || !status.kekIterations) {
    throw new Error("Vault not set up. Set up your vault in the web dashboard first.");
  }

  const salt = fromBase64(status.kekSalt);
  return deriveKEK(password, salt, status.kekIterations);
}

async function resolveDekFromKek(projectId: string, kek: CryptoKey): Promise<CryptoKey> {
  const client = getApiClient();
  const { data: grant, error } = await client.api.projects({ id: projectId }).keygrants.my.get();

  if (grant) {
    const dekRaw = await unwrapKey(grant.wrappedDek, grant.wrappedDekIv, grant.wrappedDekTag, kek);
    return importDEK(dekRaw);
  }

  // Check if this is a 404 (no grant) vs a real error
  const is404 = error && error.value.message?.includes("No key grant");

  if (!is404) {
    throw new Error(error.value.message ?? "Failed to fetch key grant.");
  }

  // Auto-create a SELF grant: generate DEK, wrap with KEK, store on server
  return createSelfGrant(projectId, kek);
}

/** Generate a new DEK, wrap it with the user's KEK, and POST a SELF grant. */
async function createSelfGrant(projectId: string, kek: CryptoKey): Promise<CryptoKey> {
  const creds = loadCredentials();
  if (!creds) {
    throw new Error("Not authenticated. Run /login first.");
  }

  const dek = await generateDEK();
  const dekRaw = await exportDEK(dek);
  const wrapped = await wrapKey(dekRaw, kek);

  const client = getApiClient();
  const { error } = await client.api.projects({ id: projectId }).keygrants.post({
    userId: creds.userId,
    wrappedDek: wrapped.wrapped,
    wrappedDekIv: wrapped.iv,
    wrappedDekTag: wrapped.tag,
    grantType: "SELF",
  });

  if (error) {
    const errorValue = error.value as { message?: string } | undefined;
    throw new Error(errorValue?.message ?? "Failed to create key grant for this project.");
  }

  return dek;
}

async function resolveDekFromCiToken(ciToken: string): Promise<CryptoKey> {
  const client = getApiClient();
  const { data, error } = await client.api.ci.secrets.get();

  if (error || !data) {
    throw new Error("Failed to fetch CI secrets. Check your DEPVAULT_TOKEN.");
  }

  const ciWrapKey = await deriveCIWrapKey(ciToken);
  const dekRaw = await unwrapKey(
    data.wrappedDek!,
    data.wrappedDekIv!,
    data.wrappedDekTag!,
    ciWrapKey,
  );
  return importDEK(dekRaw);
}
